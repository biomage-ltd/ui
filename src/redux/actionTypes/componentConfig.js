/**
 * Sets embedding categorical config.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;

export { LOAD_CONFIG, UPDATE_CONFIG, SAVE_CONFIG };
