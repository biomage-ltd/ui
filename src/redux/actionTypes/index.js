const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
