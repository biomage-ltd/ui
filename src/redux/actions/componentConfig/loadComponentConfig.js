import _ from 'lodash';
import { LOAD_CONFIG } from '../../actionTypes/componentConfig';
import { initialComponentConfigStates } from '../../reducers/componentConfig/initialState';

const loadComponentConfig = (experimentId, plotUuid, type) => (dispatch) => {
  dispatch({
    type: LOAD_CONFIG,
    payload: {
      experimentId,
      plotUuid,
      type,
      config: _.cloneDeep(initialComponentConfigStates[type]),
    },
  });
};

export default loadComponentConfig;
