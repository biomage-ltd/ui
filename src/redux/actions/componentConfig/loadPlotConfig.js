import _ from 'lodash';
import { LOAD_CONFIG } from '../../actionTypes/componentConfig';
import getApiEndpoint from '../../../utils/apiEndpoint';
import pushNotificationMessage from '../notifications';
import messages from '../../../components/notification/messages';
import { initialPlotConfigStates } from '../../reducers/componentConfig/initialState';

const loadPlotConfig = (experimentId, plotUuid, plotType) => async (dispatch) => {
  try {
    const response = await fetch(
      `${getApiEndpoint()}/v1/experiments/${experimentId}/plots-tables/${plotUuid}`,
    );
    if (response.ok) {
      const data = await response.json();
      dispatch({
        type: LOAD_CONFIG,
        payload: data,
      });
    } else if (response.status === 404) {
      dispatch({
        type: LOAD_CONFIG,
        payload: {
          experimentId,
          plotUuid,
          type: plotType,
          config: _.cloneDeep(initialPlotConfigStates[plotType]),
        },
      });
    } else {
      throw new Error('Server sent back different error or json conversion failed.');
    }
  } catch (e) {
    dispatch(pushNotificationMessage('error', messages.saveCellSets, 10));
  }
};

export default loadPlotConfig;
