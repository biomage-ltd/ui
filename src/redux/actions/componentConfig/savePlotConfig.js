import fetchAPI from '../../../utils/fetchAPI';
import { SAVE_CONFIG } from '../../actionTypes/componentConfig';
import getApiEndpoint from '../../../utils/apiEndpoint';

const savePlotConfig = (experimentId, plotUuid) => async (dispatch, getState) => {
  // Do not save the 'outstandingChanges' state to the database.
  // Do not save the 'plotData' state to the database because it is not managed by the UI.
  const { outstandingChanges, plotData, ...content } = getState().componentConfig[plotUuid];

  const response = await fetchAPI(
    `${getApiEndpoint()}/v1/experiments/${experimentId}/plots-tables/${plotUuid}`,
    {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(content),
    },
  );
  const { lastUpdated } = await response.json();

  dispatch({
    type: SAVE_CONFIG,
    payload:
      { plotUuid, lastUpdated },
  });
};

export default savePlotConfig;
