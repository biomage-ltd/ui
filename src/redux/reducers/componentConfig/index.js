import { UPDATE_CONFIG, SAVE_CONFIG, LOAD_CONFIG } from '../../actionTypes/componentConfig';
import initialState from './initialState';
import loadConfig from './loadConfig';
import updateConfig from './updateConfig';
import saveConfig from './saveConfig';

const plotsReducer = (state = initialState, action) => {
  switch (action.type) {
    case LOAD_CONFIG:
      return loadConfig(state, action);
    case UPDATE_CONFIG:
      return updateConfig(state, action);
    case SAVE_CONFIG:
      return saveConfig(state, action);
    default:
      return state;
  }
};

export default plotsReducer;
