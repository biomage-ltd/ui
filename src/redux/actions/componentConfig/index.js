import loadPlotConfig from './loadPlotConfig';
import loadComponentConfig from './loadComponentConfig';
import updatePlotConfig from './updatePlotConfig';
import savePlotConfig from './savePlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
};
