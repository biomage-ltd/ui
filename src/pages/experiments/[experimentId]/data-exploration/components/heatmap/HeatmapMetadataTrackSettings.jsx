import React, { useEffect, useState, useRef } from 'react';
import _ from 'lodash';
import { useSelector, useDispatch } from 'react-redux';
import {
  EyeOutlined,
  EyeInvisibleOutlined,
  UpOutlined,
  DownOutlined,
} from '@ant-design/icons';
import {
  Button, Space, Switch,
} from 'antd';
import { updatePlotConfig } from '../../../../../../redux/actions/componentConfig';

const HeatmapMetadataTrackSettings = () => {
  const dispatch = useDispatch();

  const cellSets = useSelector((state) => state.cellSets);
  const selectedTracks = useSelector(
    (state) => state.componentConfig.interactiveHeatmap.config.selectedTracks,
  );

  const getCellSets = (category) => {
    if (!cellSets || cellSets.loading) {
      return [];
    }

    return cellSets.hierarchy.map(
      ({ key }) => (
        { key, name: cellSets.properties[key].name, type: cellSets.properties[key].type }
      ),
    ).filter(
      ({ type }) => category.includes(type),
    );
  };

  const getTrackData = () => getCellSets(
    ['cellSets', 'metadataCategorical'],
  ).map(
    (data) => ({ selected: selectedTracks.includes(data.key), key: data.key }),
  );

  const isInitialRenderRef = useRef(true);
  const [trackData, setTrackData] = useState(getTrackData());

  const getUpdatedTrackData = () => _.unionBy(
    trackData,
    getTrackData(),
    'key',
  );

  useEffect(() => {
    // Prevent initial dispatch when object appears
    if (isInitialRenderRef.current) {
      return;
    }

    setTrackData(getUpdatedTrackData());
  }, [cellSets.hierarchy]);

  useEffect(() => {
    // Prevent initial dispatch when object appears
    if (isInitialRenderRef.current) {
      return;
    }

    if (trackData.length === 0) {
      return;
    }

    dispatch(
      updatePlotConfig('interactiveHeatmap', {
        selectedTracks: trackData.filter((o) => o.selected).map((o) => o.key),
      }),
    );
  }, [trackData]);

  useEffect(() => {
    isInitialRenderRef.current = false;
  });

  const moveUp = (source, id) => {
    const index = source.findIndex((e) => e.key === id);

    const arr = [...source];

    if (index <= 0) {
      return arr;
    }

    const el = arr[index];
    arr[index] = arr[index - 1];
    arr[index - 1] = el;

    return arr;
  };

  const moveDown = (source, id) => {
    const index = source.findIndex((e) => e.key === id);

    const arr = [...source];

    if (index === -1 || index >= source.length - 1) {
      return arr;
    }

    const el = arr[index];
    arr[index] = arr[index + 1];
    arr[index + 1] = el;

    return arr;
  };

  return (
    <div style={{ padding: '5px' }}>
      <Space direction='vertical'>
        {trackData.map(({ key, selected }, i) => (
          <Space>
            <Switch
              checkedChildren={<EyeOutlined />}
              unCheckedChildren={<EyeInvisibleOutlined />}
              value={key}
              checked={selected}
              onChange={(state) => {
                const newState = [...trackData];
                newState[i].selected = state;
                setTrackData(newState);
              }}
            />

            <div>
              <Button
                size='small'
                icon={<UpOutlined />}
                shape='circle'
                disabled={i === 0}
                onClick={() => {
                  setTrackData(moveUp(trackData, key));
                }}
              />
              <Button
                size='small'
                shape='circle'
                disabled={i === trackData.length - 1}
                icon={<DownOutlined />}
                onClick={() => {
                  setTrackData(moveDown(trackData, key));
                }}
              />
            </div>

            {cellSets.properties[key].name}
          </Space>
        ))}
      </Space>
    </div>
  );
};

HeatmapMetadataTrackSettings.defaultProps = {
};

HeatmapMetadataTrackSettings.propTypes = {
};

export default HeatmapMetadataTrackSettings;
