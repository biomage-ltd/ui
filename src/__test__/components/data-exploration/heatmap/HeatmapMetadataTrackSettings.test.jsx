import React from 'react';
import { configure, mount } from 'enzyme';
import { Provider } from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Switch } from 'antd';

import { UPDATE_CONFIG } from '../../../../redux/actionTypes/componentConfig';
import HeatmapMetadataTrackSettings from '../../../../components/data-exploration/heatmap/HeatmapMetadataTrackSettings';

jest.mock('localforage');
jest.mock('../../../../components/data-exploration/heatmap/VegaHeatmap');

const mockStore = configureMockStore([thunk]);
configure({ adapter: new Adapter() });

let component;

const initialState = {
  cellSets: {
    hierarchy: [
      {
        key: 'louvain',
        children: [
          {
            key: 'louvain-0',
          },
          {
            key: 'louvain-1',
          },
          {
            key: 'louvain-2',
          },
        ],
      },
      {
        key: 'sample',
        children: [
          {
            key: 'control',
          },
          {
            key: 'treated',
          },
        ],
      },
    ],
    properties: {
      louvain: {
        type: 'cellSets',
        name: 'louvain clusters',
      },
      'louvain-0': {
        name: 'louvain 0',
        cellIds: new Set([5, 6, 7]),
      },
      'louvain-1': {
        name: 'louvain 1',
        cellIds: new Set([1, 2, 3]),
      },
      'louvain-2': {
        name: 'louvain 2',
        cellIds: new Set([9, 11, 13]),
      },
      sample: {
        type: 'metadataCategorical',
        name: 'Sample',
      },
      control: {
        name: 'control',
        cellIds: new Set([5, 6, 7]),
      },
      treated: {
        name: 'treated',
        cellIds: new Set([1, 2, 3]),
      },
    },
  },
  cellInfo: {},
  componentConfig: {
    interactiveHeatmap: {
      config: {
        groupedTracks: ['sample'],
        selectedTracks: ['louvain'],
      },
    },
  },
};

describe('HeatmapGroupBySettings', () => {
  afterEach(() => {
    component.unmount();
  });

  it('renders correctly', () => {
    const store = mockStore({
      ...initialState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapMetadataTrackSettings componentType='interactiveHeatmap' experimentId='123' />
      </Provider>,
    );

    // Should be rendered.
    expect(component.find('HeatmapMetadataTrackSettings').length).toEqual(1);

    // With four buttons.
    const buttons = component.find('Button');
    expect(buttons.length).toEqual(4);

    // With two switches.
    const switches = component.find(Switch);
    expect(switches.length).toEqual(2);

    // The first rendered button should be disabled.
    expect(buttons.at(0).props().disabled).toEqual(true);

    // The last rendered button should be disabled.
    expect(buttons.at(buttons.length - 1).props().disabled).toEqual(true);
  });

  it('order and shown tracks are changed when the switches and the buttons are pressed', () => {
    const store = mockStore({
      ...initialState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapMetadataTrackSettings componentType='interactiveHeatmap' experimentId='123' />
      </Provider>,
    );

    // Get switches and click on second one.
    const switches = component.find(Switch);
    switches.at(1).simulate('click');

    // The store should update.
    expect(store.getActions().length).toEqual(1);
    const action = store.getActions()[0];
    expect(action.type).toBe(UPDATE_CONFIG);
    expect(action).toMatchSnapshot();

    // Get buttons.
    const buttons = component.find('Button');

    // Press the down button in the first row.
    buttons.at(1).simulate('click');

    // The store should update.
    expect(store.getActions().length).toEqual(2);
    const reorderAction = store.getActions()[1];
    expect(reorderAction.type).toBe(UPDATE_CONFIG);
    expect(reorderAction).toMatchSnapshot();
  });
});
