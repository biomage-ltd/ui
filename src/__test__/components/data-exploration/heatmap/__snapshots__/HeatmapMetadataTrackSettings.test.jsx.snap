// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapGroupBySettings order and shown tracks are changed when the switches and the buttons are pressed 1`] = `
Object {
  "payload": Object {
    "configChanges": Object {
      "selectedTracks": Array [
        "louvain",
        "sample",
      ],
    },
    "plotUuid": "interactiveHeatmap",
  },
  "type": "componentConfig/update",
}
`;

exports[`HeatmapGroupBySettings order and shown tracks are changed when the switches and the buttons are pressed 2`] = `
Object {
  "payload": Object {
    "configChanges": Object {
      "selectedTracks": Array [
        "sample",
        "louvain",
      ],
    },
    "plotUuid": "interactiveHeatmap",
  },
  "type": "componentConfig/update",
}
`;
